environment.go

```diff
@@ -27,21 +27,68 @@ type PipelineEnvironment struct {
 	Services           ServiceMetaList         `json:"services"`
 	Steps              ServiceMetaList         `json:"steps"`
 	ProjectName        string                  `json:"project_name"`
+}
+
+// PipelineEnvironment stores additional data for pipelines and steps.
+type PipelineEnvironment struct {
+	Version            string
+	Machines           []Machine
+	LogSettings        LogSettings
+	Substitutions      types.MappingWithEquals
+	TempDirPath        string
+	TempDirNoAutoClean bool
+	Steps              ServiceMetaList
+	ProjectName        string
 	tempFiles          []string
 	tempPaths          map[string]string
 }
 
+// UnmarshalJSON loads a PipelineDefinition from json using the pipelineJson struct.
+func (r *PipelineEnvironment) UnmarshalJSON(data []byte) error {
+	result := PipelineEnvironment{
+		Steps:     ServiceMetaList{},
+		tempFiles: []string{},
+		tempPaths: map[string]string{},
+	}
+	storage := pipelineEnvironmentJson{}
```

Comment:
I wouldn't call this `storage` for me `storage` is stuff like hard drives or a filesystem. Why not call it `parsed_json`

Reply:
Never thought of that name, thanks!